torchaudio/prototype/datasets/bucketize_sampler.py

```diff
@@ -0,0 +1,141 @@
+from typing import Dict, Iterator, List, Optional
+
+import torch
+from torch import Tensor
+from torch.utils.data import BatchSampler
+
+
+class BucketizeBatchSampler(BatchSampler):
+    """Buketized BatchSampler for sequential data with different lengths to reduce number of paddings.
+
+    Args:
+        lengths (List[int]): The lengths of the samples in the dataset.
+        num_buckets (int): The number of buckets to split the data samples.
+        min_len (int, optional): The minimum sample lengths to keep.
+            (Default: 0)
+        max_len (int or None, optional): The maximum sample lengths to keep. Inferred if not provided.
+            (Default ``None``)
+        max_token_count (int or None, optional): The max number of tokens in one mini-batch.
+            (Default: ``None``)
+        batch_size (int or None, optional): The number of samples in one mini-batch.
+             (Default: ``None``)
+        shuffle (bool, optional): Whether to shuffle buckets for non-monotonic length sampling.
+             (Default True)
+
+    Note:
+        ``max_token_count`` and ``batch_size`` are mutually exclusive. Only one argument of the two
+        should have value.
+    """
+
+    def __init__(
+        self,
+        lengths: List[int],
+        num_buckets: int,
+        min_len: int = 0,
+        max_len: Optional[int] = None,
+        max_token_count: Optional[int] = None,
+        batch_size: Optional[int] = None,
+        shuffle: bool = True,
+    ) -> None:
+        if max_len is None:
+            max_len = max(lengths)
+
+        if not (0 <= min_len <= max_len):
+            raise AssertionError("``min_len`` should be non-negative and smaller than ``max_len``")
+        if max_token_count is not None and batch_size is not None:
+            raise AssertionError("The ``max_token_count`` and ``batch_size`` can't be both set.")
+        if max_token_count is None and batch_size is None:
+            raise AssertionError("One of ``max_token_count`` or ``batch_size`` must be set.")
+        # Filter out samples which are outside the bounds of [min_len, max_len]
+        filtered_length_idx = [(length, i) for i, length in enumerate(lengths) if min_len <= length <= max_len]
+        if len(filtered_length_idx) == 0:
+            raise AssertionError("``lengths`` cannot be empty after filtering.")
+        # sort to minimize gap when bucketizing.
+        sorted_filtered_length_idx = sorted(filtered_length_idx, key=lambda x: x[0])
+        self.lengths = [e[0] for e in sorted_filtered_length_idx]
+        self.indices = [e[1] for e in sorted_filtered_length_idx]
+        self.index2iter = {e: i for i, e in enumerate(self.indices)}
```

Comment:
hmm i might be misunderstanding the purpose. why is sorting necessary?

Reply:
Sorting the samples by the lengths will minimize the paddings. Then for the samples in each bucket, we can shuffle them to introduce some randomness with marginal zero paddings, since they have similar lengths.

Reply:
yeah, i see the purpose in minimizing padding. but, downstream, you are using `bucketize` to assign bucket indices to the samples by length — sorting the lengths beforehand doesn't influence the assignments, and shuffling would undo the sorting within each bucket anyway

anyway, i see that you've moved shuffling into `__iter__`, so it isn't an issue now. re: the first point — i think it would be cleaner to remove `index2iter` and directly bucketize `lengths` so that you don't have to remap indices

Reply:
Oh I see. My assumption is sorting also guarantees the buckets are sorted, the if one bucket has been iterated and the batch is not filled, the samples from the next bucket will have similar length.
After second thought, this can be solved by sorting the keys in self.buckets